src/main/java/rx/internal/util/ScalarSynchronousSingle.java

```diff
@@ -0,0 +1,155 @@
+     *
+     * @param scheduler the target scheduler
+     * @return the new observable
+     */
+    public Single<T> scalarScheduleOn(Scheduler scheduler) {
+        if (scheduler instanceof EventLoopsScheduler) {
+            EventLoopsScheduler es = (EventLoopsScheduler) scheduler;
+            return create(new DirectScheduledEmission<T>(es, value));
+        }
+        return create(new NormalScheduledEmission<T>(scheduler, value));
+    }
+
+    /**
+     * Optimized observeOn for scalar value observed on the EventLoopsScheduler.
+     */
+    static final class DirectScheduledEmission<T> implements OnSubscribe<T> {
+        private final EventLoopsScheduler es;
+        private final T value;
+
+        DirectScheduledEmission(EventLoopsScheduler es, T value) {
+            this.es = es;
+            this.value = value;
+        }
+
+        @Override
+        public void call(SingleSubscriber<? super T> singleSubscriber) {
+            singleSubscriber.add(es.scheduleDirect(new ScalarSynchronousSingleAction<T>(singleSubscriber, value)));
+        }
+    }
+
+    /**
+     * Emits a scalar value on a general scheduler.
+     */
+    static final class NormalScheduledEmission<T> implements OnSubscribe<T> {
+        private final Scheduler scheduler;
+        private final T value;
+
+        NormalScheduledEmission(Scheduler scheduler, T value) {
+            this.scheduler = scheduler;
+            this.value = value;
+        }
+
+        @Override
+        public void call(SingleSubscriber<? super T> singleSubscriber) {
+            Worker worker = scheduler.createWorker();
+            singleSubscriber.add(worker);
+            worker.schedule(new ScalarSynchronousSingleAction<T>(singleSubscriber, value));
+        }
+    }
+
+    /**
+     * Action that emits a single value when called.
+     */
+    static final class ScalarSynchronousSingleAction<T> implements Action0 {
+        private final SingleSubscriber<? super T> subscriber;
+        private final T value;
+
+        ScalarSynchronousSingleAction(SingleSubscriber<? super T> subscriber,
+                                      T value) {
+            this.subscriber = subscriber;
+            this.value = value;
+        }
+
+        @Override
+        public void call() {
+            try {
+                subscriber.onSuccess(value);
+            } catch (Throwable t) {
+                subscriber.onError(t);
+            }
+        }
+    }
+
+    public <R> Single<R> scalarFlatMap(final Func1<? super T, ? extends Single<? extends R>> func) {
+        return create(new OnSubscribe<R>() {
+            @Override
+            public void call(final SingleSubscriber<? super R> child) {
+
+                Single<? extends R> o = func.call(value);
+                if (o instanceof ScalarSynchronousSingle) {
+                    child.onSuccess(((ScalarSynchronousSingle<? extends R>) o).value);
+                } else {
+                    o.unsafeSubscribe(new Subscriber<R>() {
```

Comment:
You need to put this new `Subscriber` to a new variable, and call `child.add(...)` to maintain the subscription chain.